File: python/121_Best_Time_To_Buy_And_Sell_Stock.py
# Solution 01 - Exceeds the time limit

class Solution(object):
    def maxProfit(self, prices):
        
        buy_price = prices[0]
        max_profit = 0
        
        for i in range(len(prices)):
            if (prices[i] < buy_price):
                buy_price = prices[i]
                
            for j in range(i+1, len(prices)):
                if((prices[j] > buy_price) and ((prices[j] - buy_price) > max_profit)):
                    max_profit = prices[j] - buy_price
                    
        return max_profit

# Solution 02

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        buy_price = prices[0]
        max_profit = 0
        
        for i in range(len(prices)):
            
            if buy_price > prices[i]:
                buy_price = prices[i]
                
            max_profit = max(max_profit, prices[i] - buy_price)
            
        return max_profit

File: java/121_Best_Time_To_Buy_And_Sell_Stock.java
// Solution 01 - Exceeds the time limit

class Solution {
    public int maxProfit(int[] prices) {
        int buyPrice = prices[0];
        int maxProfit = 0;
        
        for (int i = 0; i < prices.length; i++) {
            if (prices[i] < buyPrice) {
                buyPrice = prices[i];
            }
            
            for (int j = i + 1; j < prices.length; j++) {
                if (prices[j] > buyPrice) {
                    if ((prices[j] - buyPrice) > maxProfit) {
                        maxProfit = prices[j] - buyPrice; 
                    }
                }
            }
        }
        
        return maxProfit;
    }
}

// Solution 02

class Solution {
    public int maxProfit(int[] prices) {
        int buyPrice = prices[0];
        int maxProfit = 0;
        
        for (int i = 0; i < prices.length; i++) {
            if (prices[i] < buyPrice) {
                buyPrice = prices[i];
            }
            
            maxProfit = Math.max(maxProfit, prices[i] - buyPrice);
        }
        
        return maxProfit;
    }
}